File: data/schemas.ts
export interface Gate {
    name: string;
    description: string;
    lines: Record<string, string>;
    crosses: string[];
    center?: string;
    across?: number;
    fear?: string;
    sexuality?: string;
    love?: string;
    business?: string;
    circuit?: string;
    subCircuit?: string;
}

export interface Channel {
    name?: string;
    description: string;
    circuit?: string;
    subCircuit?: string;
}

export interface MetaObject {
    name: string;
    description: string;
}

export interface Center extends MetaObject {
    normal: string;
    distorted: string;
}

export interface PHSBlock {
    colors: Record<string, string>;
    tones: Record<string, string>;
}

export interface GatesDatabase {
    gates: Record<string, Gate>;
    channels: Record<string, Channel>;
    centers: Record<string, Center>;
    types: Record<string, MetaObject>;
    profiles: Record<string, MetaObject>;
    authorities: Record<string, MetaObject>;
    crosses: Record<string, MetaObject>;
    diet: PHSBlock;
    motivation: PHSBlock;
    vision: PHSBlock;
    environment: PHSBlock;
    circuits?: Record<string, CircuitGroup>;
}

export interface SubCircuit {
    name: string;
    description: string;
}

export interface CircuitGroup {
    name: string;
    description: string;
    sub_circuits: Record<string, SubCircuit>;
}

export interface CircuitMapping {
    circuits: Record<string, CircuitGroup>;
    gateMapping: Record<string, string>;
    channelMapping: Record<string, string>;
}

File: data/schemas.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize)]
pub struct Gate {
    pub name: String,
    pub description: String,
    pub lines: HashMap<String, String>,
    pub crosses: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub center: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub across: Option<u8>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fear: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sexuality: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub love: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub business: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Channel {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub description: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct MetaObject {
    pub name: String,
    pub description: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Center {
    pub name: String,
    pub normal: String,
    pub distorted: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PhsBlock {
    pub colors: HashMap<String, String>,
    pub tones: HashMap<String, String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GatesDatabase {
    pub gates: HashMap<String, Gate>,
    pub channels: HashMap<String, Channel>,
    pub centers: HashMap<String, Center>,
    pub types: HashMap<String, MetaObject>,
    pub profiles: HashMap<String, MetaObject>,
    pub authorities: HashMap<String, MetaObject>,
    pub crosses: HashMap<String, MetaObject>,
    pub diet: PhsBlock,
    pub motivation: PhsBlock,
    pub vision: PhsBlock,
    pub environment: PhsBlock,
}
